### DISCUSSION TITLE
Accessing a main resource while testing in a modular environment.

### DISCUSSION THREAD

**Discussion by TomerFi** [sormuras/junit-platform-maven-plugin#70] [General]
Hey there @sormuras 

I'm trying to figure out how to read resources in a modular environment.
As you are the **modules master** and this is **my favorite maven plugin**,
I thought I'll run this by you.
😁 

I created a testing project:

- A [ReadResourceFiles class](https://github.com/TomerFi/modular-resources-playground/blob/master/src/main/java/com/example/application/ReadResourceFiles.java) for reading the content of a resource file and returning it as a String.
- A [TextFile txt](https://github.com/TomerFi/modular-resources-playground/blob/master/src/main/resources/com/example/files/Textfile.txt), the requested resource file.
- A [ReadResourceFilesTest class](https://github.com/TomerFi/modular-resources-playground/blob/master/src/test/java/com/example/application/ReadResourceFilesTest.java) for testing the application.

I have two active branches:

 - The [master branch](https://github.com/TomerFi/modular-resources-playground/tree/master) with no `module descriptors`, as seen in [this workflow run](https://github.com/TomerFi/modular-resources-playground/runs/2509847196?check_suite_focus=true), works as expected.
 - The [added-module-descriptor branch](https://github.com/TomerFi/modular-resources-playground/tree/added-module-descriptor), as the name suggests, with `module descriptors`, as seen in [this workflow run](https://github.com/TomerFi/modular-resources-playground/runs/2509875584?check_suite_focus=true), returns `null` when trying to pick up the file.

Any thoughts about how might I go at it?

**Comment by sormuras** [comment]
In-module testing is always ... fun.

Your setup reminds me of https://github.com/sormuras/junit-platform-maven-plugin/tree/master/src/it/modular-world-3-main-foo-test-foo -- which doesn't read a resource, yet. I'll give it a shot.

**Reply by sormuras** [reply]
It didn't work out of the box with `it/modular-world-3-main-foo-test-foo`. The underlying reason is that main resources are copied to `target/classes` and test resources are copied to `target/test-classes`.

### Work-around 0

Recreate resources in `test/resources`...

### Work-around 1

Declare main resources as test resources...

```xml
  <build>
    <testResources>
      <testResource>
        <directory>${project.basedir}/src/test/resources</directory>
      </testResource>
      <testResource>
        <directory>${project.basedir}/src/main/resources</directory>
      </testResource>
    </testResources>
    ...
```

### Work-around 2

Patch main resource into each test module at runtime.

```xml
          <executor>JAVA</executor>
          <javaOptions>
            <additionalOptions>
              <additionalOption>--patch-module</additionalOption>
              <additionalOption>com.example.module=${project.basedir}/src/main/resources</additionalOption>
            </additionalOptions>
          </javaOptions>
        </configuration>
```
(Not sure about module name `com.example.module` here...)

### Work-around 3

Place above command-line option into the `module-info.test` file.

**Reply by sormuras** [reply]
Added "Work-around 2" via https://github.com/sormuras/junit-platform-maven-plugin/commit/f67feb4e1258c8acfabd4e9af1ad678f62847c97 into `it/modular-world-3-main-foo-test-foo`. Should also work in cases similar to `it/modular-world-4-main-foo-test-bar`.

**Reply by TomerFi** [reply]
👏👏👏👏
This is good stuff, thank you very much!

I was hoping for a solution involving jpms and not maven, the `patch-module` solution is sweet!

I'm away from my station, I'll try work-around-3 first thing tomorrow morning!
😀

**Reply by TomerFi** [reply]
@sormuras 

None of the workarounds worked for me out-of-the-box,
And I think I have a lead on the source for the issue.

I've tested my theory with a new testing project copied from your [modular-world-3-main-foo-test-foo it project]( https://github.com/sormuras/junit-platform-maven-plugin/tree/master/src/it/modular-world-3-main-foo-test-foo).

First of all, replacing the original `module-info.java` in the test folder:

```java
open module foo {
  requires org.junit.jupiter.api;
}
```

With a `module-info.test` will result in:

```text
--add-modules
  foo

--add-opens
  foo/foo=org.junit.platform.commons

--add-reads
  foo=org.junit.jupiter.api
```

The first thing that happens is an `ExceptionInInitializerError` is thrown with the following cause:

```shell
RuntimeException: foo specified more than once to --patch-module
```

This makes sense, I think I've noticed while debugging that using `module-info.test` is doing some patching of its own.

But... that's ok!
Now, the patching is not actually needed anymore,
both `main` and `test` resources are accessible for testing.

This is where it gets weirder, you cannot access subfolders.
If you leave your code as-is, it will still fail since it tries to access the file `foo/x.txt`.

If you move the file outside of the `foo` sub folder and place it directly in the `resources` folder.
You can easily access it after modifying the test cases accordingly of course.

Any thoughts?

**Reply by sormuras** [reply]
Pointers:
- https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/lang/ClassLoader.html#getResource(java.lang.String)
- https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/lang/Module.html#getResourceAsStream(java.lang.String)

The latter explains the subfolder access issues: you need to open each of them.

```
--add-opens
  foo/foo.sub=<module that needs to reflect into package foo.sub>
```

I guess, the root package is considered to be an illegal package name, therefore a resource placed directly in `resources ` is not encapsulated.

**Reply by TomerFi** [reply]
You're right, the [getResourceAsStream javadoc](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/lang/Module.html#getResourceAsStream(java.lang.String)) does explain the root resources not being encapsulated.

I've tried treating the subfolders as a package name and open it for reflection, but to no avail.

I think there may be something more that needs to be done,
Maybe something in the syntax of the file path...
I've tried a couple of variations, can't seem to get this to work.

I've shared the project to [github](https://github.com/TomerFi/modular-world-3-main-foo-test-foo).

**Reply by TomerFi** [reply]
So... I've made progress.
Opening `foo.sub` package for reflection in the source `module-info.class` works well:

```java
module foo {
  opens foo.sub;
}
```

But when opening `foo.sub` package within `module-info.test`,
the resource files are still not accessible:

```java
--add-opens
  foo/foo.sub=org.junit.jupiter.api
```

I guess I've got the wrong reflecting module.
It's now clear to me how was it working when using `module-info.java` with an `open module` statement.

I'll keep investigating...
I just need to locate the module accessing the resources files.
